src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1161,35 +1160,49 @@ private ImmutableArray<BoundNode> ConvertCollectionExpressionElements(
                                 implicitReceiver,
                                 static (binder, syntax, item, implicitReceiver, arg, diagnostics) =>
                                 {
-                                    Debug.Assert(ConversionsBase.IsKeyValuePairType(binder.Compilation, item.Type, out _, out _));
+                                    BoundExpression convertedExpression;
+                                    if (arg.elementConversion.TryGetKeyValueConversions(out var keyConversion, out var valueConversion))
+                                    {
+                                        Debug.Assert(ConversionsBase.IsKeyValuePairType(binder.Compilation, item.Type, out _, out _));
+                                        convertedExpression = item;
+                                    }
+                                    else
+                                    {
+                                        convertedExpression = binder.CreateConversion(item, arg.elementConversion, arg.elementType, diagnostics);
```

Comment:
It looks like in the "non-KeyValuePair" case, we can process the conversion of the element using the ordinary CreateConversion, passing along the corresponding conversion that we got from the previous phase, but, in the "KeyValuePair" case, we need to instead "deconstruct" the conversion, and pass its data along to BindDictionaryItemAssignment? Is this easier than just always creating the conversion, and having `BindDictionaryItemAssignment` pick the pieces out, in the case it is a `KeyValuePair` conversion?

Reply:
Passing along the conversion rather than deconstructing it may be simpler, although let's consider that in a subsequent change, to avoid churn for now.